dpar/src/models/tensorflow/guide.rs

```diff
@@ -0,0 +1,46 @@
+use enum_map::EnumMap;
+use tensorflow::Tensor;
+
+use system::{ParserState, TransitionSystem};
+
+use guide::{BatchGuide, Guide};
+
+use super::{LayerTensors, TensorflowModel};
+
+impl<T> Guide for TensorflowModel<T>
+where
+    T: TransitionSystem,
+{
+    type Transition = T::Transition;
+
+    fn best_transition(&mut self, state: &ParserState) -> Self::Transition {
+        self.best_transitions(&[state]).remove(0)
+    }
+}
+
+impl<T> BatchGuide for TensorflowModel<T>
+where
+    T: TransitionSystem,
+{
+    type Transition = T::Transition;
+
+    fn best_transitions(&mut self, states: &[&ParserState]) -> Vec<Self::Transition> {
+        if states.is_empty() {
+            return Vec::new();
+        }
+
+        // Allocate batch tensors.
+        let mut input_tensors = LayerTensors(EnumMap::new());
+        for (layer, size) in self.vectorizer().layer_sizes() {
+            input_tensors[layer] = Tensor::new(&[states.len() as u64, size as u64]).into();
```

Comment:
Compiler optimization might already catch that, states.len() is cast to u64 in every iteration. It could be assigned to a u64 before the loop.

Reply:
On x86_64, `usize` and `u64` have the same representation, so the cast does not exist after compilation. If you use Tensorflow on i386, you have other performance problems than this cast ;).

Also, `states.len()` will be hoisted out of the loop by the compiler - it's guaranteed that the length does not change, since `states` is an immutable reference.

Reply:
Just for fun I looked at the assembly and it seems that the loop is completely removed. The compiler realizes that the `EnumMap` is a fixed-width array and just unrolls the complete loop.